python/taichi/lang/kernel_arguments.py

```diff
@@ -30,52 +30,51 @@ def ext_arr():
 class ArgAnyArray:
     """Type annotation for arbitrary arrays, including external arrays and Taichi ndarrays.
 
-    For external arrays, we can treat it as a Taichi field with Vector or Matrix elements by specifying element shape and layout.
-    For Taichi vector/matrix ndarrays, we need to specify element shape and layout for type checking.
+    For external arrays, we can treat it as a Taichi field with Vector or Matrix elements by specifying element dim and layout.
+    For Taichi vector/matrix ndarrays, we will automatically identify element dim and layout. If they are explicitly specified, we will check compatibility between the actual arguments and the annotation.
 
     Args:
-        element_shape (Tuple[Int], optional): () if scalar elements (default), (n) if vector elements, and (n, m) if matrix elements.
-        layout (Layout, optional): Memory layout, AOS by default.
+        element_dim (Union[Int, NoneType], optional): None if not specified (will be treated as 0 for external arrays), 0 if scalar elements, 1 if vector elements, and 2 if matrix elements.
+        layout (Union[Layout, NoneType], optional): None if not specified (will be treated as Layout.AOS for external arrays), Layout.AOS or Layout.SOA.
     """
-    def __init__(self, element_shape=(), layout=Layout.AOS):
-        if len(element_shape) > 2:
+    def __init__(self, element_dim=None, layout=None):
+        if element_dim is not None and (element_dim < 0 or element_dim > 2):
             raise ValueError(
                 "Only scalars, vectors, and matrices are allowed as elements of ti.any_arr()"
             )
-        self.element_shape = element_shape
+        self.element_dim = element_dim
         self.layout = layout
 
     def extract(self, x):
         from taichi.lang.matrix import MatrixNdarray, VectorNdarray
-        from taichi.lang.ndarray import Ndarray, ScalarNdarray
-        element_dim = len(self.element_shape)
-        if isinstance(x, Ndarray):
-            if isinstance(x, ScalarNdarray) and element_dim != 0:
+        from taichi.lang.ndarray import ScalarNdarray
+        if isinstance(x, ScalarNdarray):
+            if self.element_dim is not None and self.element_dim != 0:
                 raise ValueError("Invalid argument passed to ti.any_arr()")
-            if isinstance(x, VectorNdarray) and (element_dim != 1 or
-                                                 self.element_shape[0] != x.n
-                                                 or self.layout != x.layout):
+            return x.dtype, len(x.shape), (), Layout.AOS
+        if isinstance(x, VectorNdarray):
+            if self.element_dim is not None and self.element_dim != 1:
```

Comment:
nit: as our semantic here varies case by case, maybe it's helpful to put more hints in the error message? E.g `Invalid argument passed to ti.any_arr(). This is a VectorNdarray but got element_dim = {}`. Similar improvements can be applied to layout check as well. 
Given it's repeated three times for scalar/vector/matrix it might also worth adding a helper function to do element_dim check ;)

Reply:
Helpful suggestion! Done.